euphonic/spectra.py

```diff
@@ -1656,6 +1780,149 @@ def from_dict(cls: Type[T], d: Dict[str, Any]) -> T:
                    metadata=d['metadata'])
 
 
+class Spectrum2DCollection(SpectrumCollectionMixin,
+                           Spectrum,
+                           collections.abc.Sequence):
+    """A collection of Spectrum2D with common x_data, y_data and x_tick_labels
+
+    Intended for convenient storage of contributions to spectral maps such as
+    S(Q,w). This object can be indexed or iterated to obtain individual
+    Spectrum2D.
+
+    Attributes
+    ----------
+    x_data
+        Shape (n_x_data,) or (n_x_data + 1,) float Quantity. The x_data
+        points (if size == (n_x_data,)) or x_data bin edges (if size
+        == (n_x_data + 1,))
+    y_data
+        Shape (n_y_data,) or (n_y_data + 1,) float Quantity. The y_data
+        points (if size == (n_y_data,)) or y_data bin edges (if size
+        == (n_y_data + 1,))
+    z_data
+        Shape (n_entries, n_x_data, n_y_data) float Quantity. The spectral data
+        in x and y, indexed over components
+    x_tick_labels
+        Sequence[Tuple[int, str]] or None. Special tick labels e.g. for
+        high-symmetry points. The int refers to the index in x_data the
+        label should be applied to
+    metadata
+        Dict[str, Union[int, str, LineData]] or None. Contains metadata
+        about the spectra. Keys should be strings and values should be
+        strings or integers.
+        There are some functional keys:
+
+          - 'line_data' : LineData
+                          This is a Sequence[Dict[str, Union[int, str]],
+                          it contains metadata for each spectrum in
+                          the collection, and must be of length
+                          n_entries
+    """
+
+    # Private attributes used by SpectrumCollectionMixin
+    _bin_axes = ("x", "y")
+    _spectrum_axis = "z"
+    _item_type = Spectrum2D
+
+    def __init__(
+            self, x_data: Quantity, y_data: Quantity, z_data: Quantity,
+            x_tick_labels: Optional[Sequence[Tuple[int, str]]] = None,
+            metadata: Optional[Metadata] = None
+    ) -> None:
+        _check_constructor_inputs(
+            [z_data, x_tick_labels, metadata],
+            [Quantity, [list, type(None)], [dict, type(None)]],
+            [(-1, -1, -1), (), ()],
+            ['z_data', 'x_tick_labels', 'metadata'])
+        nx = z_data.shape[1]
+        ny = z_data.shape[2]
```

Comment:
Just to check, these shouldn't be `[0]` and `[1]`.

If not, would it be better as:
```python
_, nx, ny = z_data.shape
```

Reply:
Always just prefer to check as someone who has to switch languages regularly from 0-indexed and 1-indexed and back again.